File: CS/Calendar_CellStyleProvider/Form1.cs
using DevExpress.Utils;
using DevExpress.Utils.Design;
using DevExpress.XtraEditors.Controls;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Calendar_CellStyleProvider {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e) {
            dateEdit1.DateTime = calendarControl1.DateTime = new DateTime(2016, 12, 31);
            dateEdit1.Properties.CellStyleProvider = calendarControl1.CellStyleProvider = new CustomCellStyleProvider();

            ContextButton cb = new ContextButton() {
                Alignment = ContextItemAlignment.TopNear, Visibility=ContextItemVisibility.Hidden

            };
            calendarControl1.CellSize = new Size(50, 50);
            calendarControl1.ContextButtons.Add(cb);
            calendarControl1.ContextButtonCustomize += CalendarControl1_ContextButtonCustomize;
        }   

        private void CalendarControl1_ContextButtonCustomize(object sender, CalendarContextButtonCustomizeEventArgs e) {
            string holidayText;
            if (Holidays.IsHoliday(e.Cell.Date, out holidayText)) {
                e.Item.Glyph = global::Calendar_CellStyleProvider.Properties.Resources.Party;
                e.Item.Visibility = ContextItemVisibility.Visible;
                e.Item.ToolTip = holidayText;
                e.Item.ShowToolTips = true;
            }
        }
    }


    public static class Holidays {
        public static bool IsHoliday(DateTime dt, out string holidayText) {
            holidayText = "";
            //New Year's Day
            if (dt.Day == 1 && dt.Month == 1) holidayText = "New Year's Day";
            //Independence Day
            if (dt.Day == 4 && dt.Month == 7) holidayText = "Independence Day";
            //Veterans Day
            if (dt.Day == 11 && dt.Month == 11) holidayText = "Veterans Day";
            //Christmas
            if (dt.Day == 25 && dt.Month == 12) holidayText = "Christmas";
            return !string.IsNullOrEmpty(holidayText);
        }
    }

    public class CustomCellStyleProvider : ICalendarCellStyleProvider {
        static Font font;
        public void UpdateAppearance(CalendarCellStyle cell) {
            string holidayText;
            if(Holidays.IsHoliday(cell.Date, out holidayText)) {
                cell.Appearance.ForeColor = Color.Yellow;
                if(font == null) {
                    font = new Font(cell.Appearance.Font, FontStyle.Bold);
                }
                cell.Appearance.Font = font;
                if (cell.Active)
                    cell.Appearance.BackColor = Color.HotPink;
                else
                    cell.Appearance.BackColor = Color.LightPink;
            }
        }
    }
}

File: VB/Calendar_CellStyleProvider/Form1.vb
Imports DevExpress.Utils
Imports DevExpress.Utils.Design
Imports DevExpress.XtraEditors.Controls
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace Calendar_CellStyleProvider
    Partial Public Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            calendarControl1.DateTime = New Date(2016, 12, 31)
            dateEdit1.DateTime = calendarControl1.DateTime
            calendarControl1.CellStyleProvider = New CustomCellStyleProvider()
            dateEdit1.Properties.CellStyleProvider = calendarControl1.CellStyleProvider

            Dim cb As New ContextButton() With {.Alignment = ContextItemAlignment.TopNear, .Visibility=ContextItemVisibility.Hidden}
            calendarControl1.CellSize = New Size(50, 50)
            calendarControl1.ContextButtons.Add(cb)
            AddHandler calendarControl1.ContextButtonCustomize, AddressOf CalendarControl1_ContextButtonCustomize
        End Sub

        Private Sub CalendarControl1_ContextButtonCustomize(ByVal sender As Object, ByVal e As CalendarContextButtonCustomizeEventArgs)
            Dim holidayText As String = Nothing
            If Holidays.IsHoliday(e.Cell.Date, holidayText) Then
                e.Item.Glyph = My.Resources.Party
                e.Item.Visibility = ContextItemVisibility.Visible
                e.Item.ToolTip = holidayText
                e.Item.ShowToolTips = True
            End If
        End Sub
    End Class


    Public NotInheritable Class Holidays

        Private Sub New()
        End Sub

        Public Shared Function IsHoliday(ByVal dt As Date, <System.Runtime.InteropServices.Out()> ByRef holidayText As String) As Boolean
            holidayText = ""
            'New Year's Day
            If dt.Day = 1 AndAlso dt.Month = 1 Then
                holidayText = "New Year's Day"
            End If
            'Independence Day
            If dt.Day = 4 AndAlso dt.Month = 7 Then
                holidayText = "Independence Day"
            End If
            'Veterans Day
            If dt.Day = 11 AndAlso dt.Month = 11 Then
                holidayText = "Veterans Day"
            End If
            'Christmas
            If dt.Day = 25 AndAlso dt.Month = 12 Then
                holidayText = "Christmas"
            End If
            Return Not String.IsNullOrEmpty(holidayText)
        End Function
    End Class

    Public Class CustomCellStyleProvider
        Implements ICalendarCellStyleProvider

        Shared font As Font
        Public Sub UpdateAppearance(ByVal cell As CalendarCellStyle) Implements ICalendarCellStyleProvider.UpdateAppearance
            Dim holidayText As String = Nothing
            If Holidays.IsHoliday(cell.Date, holidayText) Then
                cell.Appearance.ForeColor = Color.Yellow
                If font Is Nothing Then
                    font = New Font(cell.Appearance.Font, FontStyle.Bold)
                End If
                cell.Appearance.Font = font
                If cell.Active Then
                    cell.Appearance.BackColor = Color.HotPink
                Else
                    cell.Appearance.BackColor = Color.LightPink
                End If
            End If
        End Sub
    End Class
End Namespace
